documentation/adr/0021-use-keyword-args.md

```diff
@@ -0,0 +1,73 @@
+[<-previous](./0020-use-c-mex-api.md) |
+next->
+
+# 21 - Use keyword args in Python and MATLAB
+
+Date: 2021-Feb-16
+
+## Status
+
+Proposal
+
+## Context
+
+### Python 
+
+Currently, we don’t have an established syntax for Python. Because we have a clean slate it makes sense to only have a Pythonic syntax using required positional, optional positional, and optional keyword/value pairs
+
+```python
+def function_name(a, b, *args, **kwargs):
+	# body
+```
+
+ e.g. `function_name(a, b, c, d, keyword_1=val_1, keyword_2=val_2, …)`
+
+Python does not natively support flag arguments, and they are not used in the core language. They are supported in a small number of Python libraries, notably `matplotlib`, which was modelled on the MATLAB API, and may be implemented on any  function as optional string arguments with custom handling.
+
+
+### Matlab 
+
+The Horace/Herbert source uses a wide range of API syntax:
+
+- `flagname` and `-flagname` flag arguments
+- true and false flags: `flagname` (`true`) and `noflagname` (`false`)
+- key-value pairs `'keyword', value, ...`
+- positional parameters
+- optional parameters interspersed with positional parameters in some functions
+
+The key-value pair syntax is equivalent to the Pythonic syntax, e.g. `function_name(a, b, c, d, ‘keywordA’, valA, ‘keywordB’, true, …)`.
+
+The mix of `-flagname` and `flagname` introduces complexity into argument parsing - specifically identifying if a string argument `flagname` a flag or a keyword argument?
+
+
+## Decision
+
+The MATLAB APIs will support positional, optional flag and keyword arguments:
+
+- flag arguments will be prefixed with `-`
+- if the same parameter is passed as both a flag and keyword argument an error will be raised if their values are inconsistent
```

Comment:
Define behaviour otherwise (i.e. OK)
e.g.
`function(a, b, '-flag', 'flag', true)`

Reply:
if they are consistent there is nothing to define: the flag will take the value assigned via both syntaxes

Reply:
Exactly. Do you think it's worth mentioning?